File: system-tests/fixtures/api-impl-native/ReactNativeErnmovieApiImplNative/ios/ElectrodeApiImpl/ElectrodeReactNativeBridge/ElectrodeBridgeTransceiver.m
/*
 * Copyright 2017 WalmartLabs
 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 
 * http://www.apache.org/licenses/LICENSE-2.0
 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "ElectrodeBridgeTransceiver.h"
#import "ElectrodeBridgeTransceiver_Internal.h"
#import "ElectrodeEventDispatcher.h"
#import "ElectrodeRequestDispatcher.h"
#import "ElectrodeBridgeTransaction.h"
#import "ElectrodeEventRegistrar.h"
#import "ElectrodeRequestRegistrar.h"
#import "ElectrodeLogger.h"

#if __has_include(<React/RCTLog.h>)
#import <React/RCTLog.h>
#elif __has_include("RCTLog.h")
#import "RCTLog.h"
#else
#import "React/RCTLog.h" // Required when used as a Pod in a Swift project
#endif

#if __has_include(<React/RCTBridge.h>)
#import <React/RCTBridge.h>
#elif __has_include("RCTBridge.h")
#import "RCTBridge.h"
#else
#import "React/RCTBridge.h" // Required when used as a Pod in a Swift project
#endif

#import "ElectrodeBridgeMessage.h"
#import "ElectrodeBridgeEvent.h"

NS_ASSUME_NONNULL_BEGIN

@interface ElectrodeBridgeTransceiver ()

@property(nonatomic, copy) NSString *name;
@property(nonatomic, strong) ElectrodeEventDispatcher *eventDispatcher;
@property(nonatomic, strong) ElectrodeRequestDispatcher *requestDispatcher;
@property(nonatomic, copy) NSMutableDictionary<NSString *, ElectrodeBridgeTransaction *> *pendingTransaction;
@property(nonatomic, assign) dispatch_queue_t syncQueue; // this is used to make sure access to
                                // pendingTransaction is thread safe.

@end

static dispatch_once_t onceToken;
static ElectrodeRequestRegistrar *requestRegistrar;
static ElectrodeEventRegistrar *eventRegistrar;
static ElectrodeRequestDispatcher *requestDispatcher;
static ElectrodeEventDispatcher *eventDispatcher;
static NSMutableDictionary *pendingTransaction;
static NSMutableArray<id<ConstantsProvider>> *constantsProviders;

@implementation ElectrodeBridgeTransceiver

+ (instancetype)sharedInstance {
  return sharedInstance;
}

- (instancetype)init {
  if (self = [super init]) {
    dispatch_once(&onceToken, ^{
      requestRegistrar = [[ElectrodeRequestRegistrar alloc] init];
      eventRegistrar = [[ElectrodeEventRegistrar alloc] init];
      requestDispatcher = [[ElectrodeRequestDispatcher alloc]
          initWithRequestRegistrar:requestRegistrar];
      eventDispatcher = [[ElectrodeEventDispatcher alloc]
          initWithEventRegistrar:eventRegistrar];
      pendingTransaction = [[NSMutableDictionary alloc] init];
      constantsProviders = [[NSMutableArray alloc] init];
    });

    _requestDispatcher = requestDispatcher;
    _eventDispatcher = eventDispatcher;
    _pendingTransaction = pendingTransaction;
  }
  return self;
}
RCT_EXPORT_MODULE(ElectrodeBridge);
+ (NSArray *)electrodeModules {
  return @[ [[ElectrodeBridgeTransceiver alloc] init] ];
}

- (NSArray *)supportedEvents {
  return @[ @"electrode.bridge.message" ];
}

#pragma ConstantsProvider implementation

- (NSDictionary<NSString *, id> *)constantsToExport {
  NSMutableDictionary<NSString *, id> *constants = [NSMutableDictionary new];
  if (constantsProviders != nil && [constantsProviders count] > 0) {
    for (id<ConstantsProvider> constant in constantsProviders) {
      [constants addEntriesFromDictionary:[constant constantsToExport]];
    }
    return constants;
  }
  ERNDebug(@"Constants provider is empty %@", constantsProviders);
  return nil;
}

- (void)addConstantsProvider:(id<ConstantsProvider>)constantsProvider {
  @synchronized(self) {
    [constantsProviders addObject:constantsProvider];
  }
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////
#pragma ElectrodeNativeBridge implementation
- (void)sendRequest:(ElectrodeBridgeRequest *)request
    completionHandler:(ElectrodeBridgeResponseCompletionHandler)completion {
  [self handleRequest:request completionHandler:completion];
}

- (void) registerRequestCompletionHandlerWithName:(NSString *)name
                                             uuid: (NSUUID *) uuid
                                       completion: (ElectrodeBridgeRequestCompletionHandler) completion {
    [self.requestDispatcher.requestRegistrar registerRequestCompletionHandlerWithName:name
                                                                                 uuid:uuid completion:completion];
}


- (nullable ElectrodeBridgeRequestCompletionHandler) unregisterRequestHandlerWithUUID:(NSUUID *)uuid {
    return [requestRegistrar unregisterRequestHandler:uuid];
}

- (void)resetRegistrar {
  [self.requestDispatcher.requestRegistrar reset];
}

- (void)sendEvent:(ElectrodeBridgeEvent *)event {
  ERNDebug(@"ElectrodeBridgeTransceiver: emit event named: %@, id: %@",
           event.name, event.messageId);
  [self notifyNativeEventListenerWithEvent:event];
  [self notifyReactNativeEventListenerWithEvent:event];
}

- (void) registerEventListenerWithName: (NSString *_Nonnull)name
                                  uuid: (NSUUID * _Nonnull)uuid
                              listener: (ElectrodeBridgeEventListener _Nonnull)eventListener {
  ERNDebug(@"%@, Adding eventListener %@ for event %@",
           NSStringFromClass([self class]), eventListener, name);
  [self.eventDispatcher.eventRegistrar registerEventListener:eventListener
                                                          name:name uuid:uuid];
}
- (nullable ElectrodeBridgeEventListener)removeEventListnerWithUUID: (NSUUID *) uuid {
    ERNDebug(@"Removing event listener with NNUUID with string %@", uuid.UUIDString);
    return [eventRegistrar unregisterEventListener:uuid];
}
#pragma ElectrodeReactBridge

RCT_EXPORT_METHOD(sendMessage : (NSDictionary *)bridgeMessage) {
  ERNDebug(@"Received message from JS(data=%@)", bridgeMessage);
  NSString *typeString =
      (NSString *)[bridgeMessage objectForKey:kElectrodeBridgeMessageType];
  ElectrodeMessageType type =
      [ElectrodeBridgeMessage typeFromString:typeString];
  switch (type) {
  case ElectrodeMessageTypeRequest: {
    ElectrodeBridgeRequest *request =
        [ElectrodeBridgeRequest createRequestWithData:bridgeMessage];
    [self handleRequest:request completionHandler:nil];
    break;
  }

  case ElectrodeMessageTypeResponse: {
    ElectrodeBridgeResponse *response =
        [ElectrodeBridgeResponse createResponseWithData:bridgeMessage];
    if (response != nil) {
      [self handleResponse:response];
    } else {
      [NSException raise:@"invalue resonse"
                  format:@"cannot construct a response from data"];
    }
    break;
  }
  case ElectrodeMessageTypeEvent: {
    ElectrodeBridgeEvent *event =
        [ElectrodeBridgeEvent createEventWithData:bridgeMessage];
    if (event != nil) {
      [self notifyNativeEventListenerWithEvent:event];
    } else {
      [NSException raise:@"invalue event"
                  format:@"cannot construct an event from data"];
    }
    break;
  }
  case ElectrodeMessageTypeUnknown:
  default:
    [NSException raise:@"invalue message"
                format:@"cannot construct any message from data"];
    break;
  }
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////
#pragma private methods

- (void)emitMessage:(ElectrodeBridgeMessage *_Nonnull)bridgeMessage {
  ERNDebug(@"Sending bridgeMessage(%@) to JS", bridgeMessage);
  [self sendEventWithName:@"electrode.bridge.message"
                     body:[bridgeMessage toDictionary]];
}

- (void)notifyReactNativeEventListenerWithEvent:(ElectrodeBridgeEvent *)event {
  [self emitMessage:event];
}

- (void)notifyNativeEventListenerWithEvent:(ElectrodeBridgeEvent *)event {
  [self.eventDispatcher dispatchEvent:event];
}

- (void)handleRequest:(ElectrodeBridgeRequest *)request
    completionHandler:
        (ElectrodeBridgeResponseCompletionHandler _Nullable)completion {
  [self logRequest:request];

  if (completion == nil && !request.isJsInitiated) {
    [NSException
         raise:@"invalid operation"
        format:@"A response lister is required for a native initiated request"];
  }

  ElectrodeBridgeTransaction *transaction =
      [self createTransactionWithRequest:request completionHandler:completion];
  if ([self.requestDispatcher canHandlerRequestWithName:request.name]) {
    [self dispatchRequestToNativeHandlerForTransaction:transaction];
  } else if (!request.isJsInitiated) { // GOTCHA: Make sure not send a request
                                       // back to JS if it's initiated on JS
                                       // side
    [self dispatchRequestToReactHandlerForTransaction:transaction];
  } else {
    ERNDebug(@"No handler available to handle request(%@)", request);
    id<ElectrodeFailureMessage> failureMessage = [ElectrodeBridgeFailureMessage
        createFailureMessageWithCode:@"ENOHANDLER"
                             message:@"No registered request handler found"];
    ElectrodeBridgeResponse *response =
        [ElectrodeBridgeResponse createResponseForRequest:request
                                         withResponseData:nil
                                       withFailureMessage:failureMessage];
    [self handleResponse:response];
  }
}

- (ElectrodeBridgeTransaction *)
createTransactionWithRequest:(ElectrodeBridgeRequest *)request
           completionHandler:
               (ElectrodeBridgeResponseCompletionHandler)completion {
  ElectrodeBridgeTransaction *transaction =
      [[ElectrodeBridgeTransaction alloc] initWithRequest:request
                                        completionHandler:completion];

  @synchronized(self) {
    [self.pendingTransaction setObject:transaction forKey:request.messageId];
    if ([request timeoutMs] != kElectrodeBridgeRequestNoTimeOut) {
      [self startTimeOutCheckForTransaction:transaction];
    }
  }

  return transaction;
}

- (void)startTimeOutCheckForTransaction:
    (ElectrodeBridgeTransaction *)transaction {
  // Add the timeout handler
  __weak ElectrodeBridgeTransceiver *weakSelf = self;
  dispatch_after(
      dispatch_time(DISPATCH_TIME_NOW,
                    (int64_t)(transaction.request.timeoutMs * NSEC_PER_MSEC)),
      dispatch_get_main_queue(), ^{
        ElectrodeBridgeResponseCompletionHandler completionBlock =
            transaction.completion;
        if (completionBlock) {
          id<ElectrodeFailureMessage> failureMessage =
              [ElectrodeBridgeFailureMessage
                  createFailureMessageWithCode:@"TIMEOUT"
                                       message:@"transaction timed out for "
                                               @"request"];
          ElectrodeBridgeResponse *response = [ElectrodeBridgeResponse
              createResponseForRequest:transaction.request
                      withResponseData:nil
                    withFailureMessage:failureMessage];
          [weakSelf handleResponse:response];
        } else {
          ERNDebug(
              @"Empty failure block. Time out may not be handled property");
        }
      });
}

- (void)dispatchRequestToNativeHandlerForTransaction:
    (ElectrodeBridgeTransaction *)transaction {
  ERNDebug(@"Sending request(%@) to native handler", transaction.request);
  ElectrodeBridgeRequest *request = transaction.request;
  __weak ElectrodeBridgeTransceiver *weakSelf = self;

  [self.requestDispatcher
        dispatchRequest:request
      completionHandler:^(
          id _Nullable data,
          id<ElectrodeFailureMessage> _Nullable failureMessage) {
        if (failureMessage != nil) {
          ElectrodeBridgeResponse *response =
              [ElectrodeBridgeResponse createResponseForRequest:request
                                               withResponseData:nil
                                             withFailureMessage:failureMessage];
          [weakSelf handleResponse:response];
        } else {
          ElectrodeBridgeResponse *response =
              [ElectrodeBridgeResponse createResponseForRequest:request
                                               withResponseData:data
                                             withFailureMessage:nil];
          [weakSelf handleResponse:response];
        }
      }];
}

- (void)dispatchRequestToReactHandlerForTransaction:
    (ElectrodeBridgeTransaction *)transaction {
  ERNDebug(@"Sending request(%@) over to JS handler because there is no local "
           @"request handler available",
           transaction.request);
  [self emitMessage:transaction.request];
}

- (void)handleResponse:(ElectrodeBridgeResponse *)response {
  ERNDebug(@"hanlding bridge response: %@", response);
  ElectrodeBridgeTransaction *transaction;
  @synchronized(self) {
    transaction = (ElectrodeBridgeTransaction *)[self.pendingTransaction
        objectForKey:response.messageId];
  }
  if (transaction != nil) {
    transaction.response = response;
    [self completeTransaction:transaction];
  } else {
    ERNDebug(@"Response(%@) will be ignored because the transcation for this "
             @"request has been removed from the queue. Perhaps it's already "
             @"timed-out or completed.",
             response);
  }
}

- (void)completeTransaction:(ElectrodeBridgeTransaction *)transaction {
  if (transaction.response == nil) {
    [NSException raise:@"invalid transaction"
                format:@"Cannot complete transaction, a transaction can only "
                       @"be completed with a valid response"];
  }
  ERNDebug(@"completing transaction(id=%@", transaction.transactionId);

  [self.pendingTransaction removeObjectForKey:transaction.transactionId];

  ElectrodeBridgeResponse *response = transaction.response;
  [self logResponse:response];

  if (transaction.isJsInitiated) {
    ERNDebug(@"Completing transaction by emitting event back to JS since the "
             @"request is initiated from JS side");
    [self emitMessage:response];
  } else {
    if (transaction.completion == nil) {
      [NSException raise:@"invalid transaction"
                  format:@"Should never reach here. A response listener should "
                         @"always be set for a local transaction"];
    } else {
      if (response.failureMessage != nil) {

        ERNDebug(@"Completing transaction by issuing a failure call back to "
                 @"local response listener");
        dispatch_async(dispatch_get_main_queue(), ^{
          transaction.completion(nil, response.failureMessage);
        });
      } else {
        ERNDebug(@"Completing transaction by issuing a success call back to "
                 @"local response listener");
        dispatch_async(dispatch_get_main_queue(), ^{
          transaction.completion(response.data, nil);
        });
      }
    }
  }
}

- (void)logRequest:(ElectrodeBridgeRequest *)request {
  ERNDebug(@"--> --> --> --> --> Request(%@)", request);
}

- (void)logResponse:(ElectrodeBridgeResponse *)response {
  ERNDebug(@"<-- <-- <-- <-- <-- Response(%@)", response);
}

+ (void)registerReactNativeReadyListener:
    (ElectrodeBridgeReactNativeReadyListner)listner {
  if (isReactNativeReady) {
    if (listner) {
      listner(sharedInstance);
    }
  }

  reactNativeReadyListener = [listner copy];
}

- (void)onReactNativeInitialized {
  isReactNativeReady = YES;
  sharedInstance = self;
  if (reactNativeReadyListener) {
    reactNativeReadyListener(self);
  }
}
+ (void)registerReactTransceiverReadyListner:
    (ElectrodeBridgeReactNativeReadyListner)listener {
  if (isTransceiverReady) {
    if (listener) {
      listener(sharedInstance);
    }
  }
  reactNativeTransceiver = listener;
}

- (void)onTransceiverModuleInitialized {
  isTransceiverReady = YES;
  sharedInstance = self;
  if (reactNativeTransceiver) {
    reactNativeTransceiver(self);
  }
}

@end
NS_ASSUME_NONNULL_END

File: system-tests/fixtures/android-container/lib/src/main/java/com/walmartlabs/electrode/reactnative/bridge/ElectrodeBridgeTransceiver.java
/*
 * Copyright 2017 WalmartLabs
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.walmartlabs.electrode.reactnative.bridge;

import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.VisibleForTesting;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.walmartlabs.electrode.reactnative.bridge.helpers.Logger;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

import static com.walmartlabs.electrode.reactnative.bridge.ElectrodeBridgeRequest.NO_TIMEOUT;

/**
 * Class that is responsible for transmitting messages between native and react native
 */
class ElectrodeBridgeTransceiver extends ReactContextBaseJavaModule implements ElectrodeNativeBridge, ElectrodeReactBridge {

    private static final String TAG = ElectrodeBridgeTransceiver.class.getSimpleName();

    private final ReactContextWrapper mReactContextWrapper;

    // Singleton instance of the bridge
    private static ElectrodeBridgeTransceiver sInstance;

    private static final ConcurrentHashMap<String, BridgeTransaction> sPendingTransactions = new ConcurrentHashMap<>();
    private static final EventRegistrar<ElectrodeBridgeEventListener<ElectrodeBridgeEvent>> sEventRegistrar = new EventRegistrarImpl<>();
    private static final EventDispatcher sEventDispatcher = new EventDispatcherImpl(sEventRegistrar);
    private static final RequestRegistrar<ElectrodeBridgeRequestHandler<ElectrodeBridgeRequest, Object>> sRequestRegistrar = new RequestRegistrarImpl<>();
    private static final RequestDispatcher sRequestDispatcher = new RequestDispatcherImpl(sRequestRegistrar);
    private static final List<ConstantsProvider> sConstantsProviders = new ArrayList<>();
    private static boolean sIsReactNativeReady;

    /**
     * Initializes a new instance of ElectrodeBridgeTransceiver
     *
     * @param reactContextWrapper The react application context
     */
    private ElectrodeBridgeTransceiver(@NonNull ReactContextWrapper reactContextWrapper) {
        super(reactContextWrapper.getContext());
        mReactContextWrapper = reactContextWrapper;
    }

    /**
     * Creates the ElectrodeBridgeTransceiver singleton
     *
     * @param reactApplicationContext The react application context
     * @return The singleton instance of ElectrodeBridgeTransceiver
     */
    static ElectrodeBridgeTransceiver create(ReactApplicationContext reactApplicationContext) {
        return create(new ReactContextWrapperInternal(reactApplicationContext));
    }

    /**
     * Creates the ElectrodeBridgeTransceiver singleton
     *
     * @param reactContextWrapper {@link ReactContextWrapper}
     * @return The singleton instance of ElectrodeBridgeTransceiver
     */
    @VisibleForTesting
    static ElectrodeBridgeTransceiver create(@NonNull ReactContextWrapper reactContextWrapper) {
        Logger.d(TAG, "Creating ElectrodeBridgeTransceiver instance");
        sInstance = new ElectrodeBridgeTransceiver(reactContextWrapper);
        return sInstance;
    }

    /**
     * Returns the singleton instance of the bridge
     */
    public static ElectrodeBridgeTransceiver instance() {
        if (sInstance == null) {
            throw new IllegalStateException("Bridge transceiver instance has not been created yet. Transceiver requires a valid ReactContext to get initialized. Make sure to that BridgePackage is added to react module which normally initializes transceiver.");
        }
        return sInstance;
    }

    /**
     * @return the name of this module. This will be the name used to {@code require()} this module
     * from javascript.
     */
    @Override
    public String getName() {
        return "ElectrodeBridge";
    }

    @Nullable
    @Override
    public Map<String, Object> getConstants() {
        if (sConstantsProviders != null && !sConstantsProviders.isEmpty()) {
            Map<String, Object> constants = new HashMap<>();
            try {
                for (ConstantsProvider provider : sConstantsProviders) {
                    Map<String, Object> providerConstants;
                    if ((providerConstants = provider.getConstants()) != null) {
                        constants.putAll(providerConstants);
                    }
                }
                return constants;
            } catch (Exception e) {
                //GOTCHA: Added a try catch since the implementation of this would be on the client side and bridge has no control over unseen errors.
                Logger.w(TAG, "getConstants() implementation by(%s) failed due to(%s)", sConstantsProviders, e.getMessage());
            }
        }
        return super.getConstants();
    }

    @NonNull
    @Override
    public boolean addEventListener(@NonNull String name, @NonNull ElectrodeBridgeEventListener<ElectrodeBridgeEvent> eventListener, @NonNull UUID uuid) {
        Logger.d(TAG, "Adding eventListener(%s) for event(%s)", eventListener, name);
        return sEventRegistrar.registerEventListener(name, eventListener, uuid);
    }

    /**
     * Registers the {@link ConstantsProvider} that will be used by the bridge module to get the constant values exposed to JavaScript
     *
     * @param constantsProvider
     */
    public static void addConstantsProvider(@NonNull ConstantsProvider constantsProvider) {
        sConstantsProviders.add(constantsProvider);
    }

    @Nullable
    @Override
    public UUID getRequestHandlerId(@NonNull String name) {
        return sRequestRegistrar.getRequestHandlerId(name);
    }

    @NonNull
    @Override
    public UUID getEventListenerId(@NonNull ElectrodeBridgeEventListener<ElectrodeBridgeEvent> eventListener) {
        return sEventRegistrar.getEventListenerId(eventListener);
    }

    @Override
    public ElectrodeBridgeEventListener<ElectrodeBridgeEvent> removeEventListener(@NonNull UUID eventListenerUuid) {
        return sEventRegistrar.unregisterEventListener(eventListenerUuid);
    }

    @Override
    public ElectrodeBridgeRequestHandler<ElectrodeBridgeRequest, Object> unregisterRequestHandler(@NonNull UUID requestHandlerUuid) {
        return sRequestRegistrar.unregisterRequestHandler(requestHandlerUuid);
    }

    @Override
    public boolean registerRequestHandler(@NonNull String name, @NonNull ElectrodeBridgeRequestHandler<ElectrodeBridgeRequest, Object> requestHandler, @NonNull UUID uuid) {
        return sRequestRegistrar.registerRequestHandler(name, requestHandler, uuid);
    }

    /**
     * Emits an event with some data to the JS react native side
     *
     * @param event The event to emit
     */
    @SuppressWarnings("unused")
    @Override
    public void sendEvent(@NonNull ElectrodeBridgeEvent event) {
        Logger.d(TAG, "Emitting event[name:%s id:%s]", event.getName(), event.getId());

        notifyReactEventListeners(event);
        notifyLocalEventListeners(event);
    }


    /**
     * Sends a request
     *
     * @param request          The request to send
     * @param responseListener Listener to be called upon request completion
     */
    @SuppressWarnings("unused")
    @Override
    public void sendRequest(@NonNull final ElectrodeBridgeRequest request, @NonNull final ElectrodeBridgeResponseListener<ElectrodeBridgeResponse> responseListener) {
        handleRequest(request, responseListener);
    }

    /**
     * This method is used by react native to dispatch an event on the native side.
     * <p>
     * This could be a REQUEST, RESPONSE, or an EVENT
     *
     * @param data The event data
     */
    @ReactMethod
    @Override
    public void sendMessage(@NonNull final ReadableMap data) {
        Logger.d(TAG, "received message from JS(data=%s)", data);
        BridgeMessage.Type type = BridgeMessage.Type.getType(data.getString(BridgeMessage.BRIDGE_MSG_TYPE));
        if (type != null) {
            switch (type) {
                case EVENT:
                    ElectrodeBridgeEvent event = ElectrodeBridgeEvent.create(data);
                    if (event != null) {
                        Logger.d(TAG, "Received message is an EVENT(name=%s), will notify local event listeners.", event.getName());
                        notifyLocalEventListeners(event);
                    } else {
                        throw new IllegalArgumentException("Unable to construct event from data");
                    }
                    break;
                case REQUEST:
                    ElectrodeBridgeRequest request = ElectrodeBridgeRequest.create(data);
                    if (request != null) {
                        Logger.d(TAG, "Received message is a REQUEST(name=%s), will look for a request handler and forward this request", request.getName());
                        handleRequest(request, null);
                    } else {
                        throw new IllegalArgumentException("Unable to construct request from data");
                    }

                    break;
                case RESPONSE:
                    ElectrodeBridgeResponse response = ElectrodeBridgeResponse.create(data);
                    if (response != null) {
                        Logger.d(TAG, "Received message is a RESPONSE for a request(name=%s, id=%s)", response.getName(), response.getId());
                        handleResponse(response);
                    } else {
                        throw new IllegalArgumentException("Unable to construct a response from data");
                    }

                    break;
            }
        } else {
            throw new IllegalArgumentException("Unable to identify request type. Should never reach here.");
        }
    }

    private void handleRequest(@NonNull final ElectrodeBridgeRequest request, @Nullable ElectrodeBridgeResponseListener<ElectrodeBridgeResponse> responseListener) {
        logRequest(request);

        if (responseListener == null && !request.isJsInitiated()) {
            throw new IllegalArgumentException("A response lister is required for a non-JS initiated request");
        }

        final BridgeTransaction bridgeTransaction = createTransaction(request, responseListener);

        if (sRequestDispatcher.canHandleRequest(request.getName())) {
            dispatchRequestToLocalHandler(bridgeTransaction);
        } else if (!request.isJsInitiated()) {//GOTCHA: Should not send a request back JS if it was initiated from JS side.
            dispatchRequestToReact(bridgeTransaction);
        } else {
            Logger.d(TAG, "No handler available to handle the request(id=%s, name=%s). Will fail the request", request.getId(), request.getName());
            handleResponse(ElectrodeBridgeResponse.createResponseForRequest(request, null, BridgeFailureMessage.create("ENOHANDLER", "No registered request handler found for " + request.getName())));
        }
    }

    @NonNull
    private BridgeTransaction createTransaction(@NonNull ElectrodeBridgeRequest request, @Nullable ElectrodeBridgeResponseListener<ElectrodeBridgeResponse> responseListener) {
        final BridgeTransaction bridgeTransaction = new BridgeTransaction(request, responseListener);
        sPendingTransactions.put(request.getId(), bridgeTransaction);
        startTimeOutCheckForTransaction(bridgeTransaction);
        return bridgeTransaction;
    }

    private void startTimeOutCheckForTransaction(@NonNull final BridgeTransaction transaction) {
        if (transaction.getRequest().getTimeoutMs() != NO_TIMEOUT) {
            Handler handler = new Handler(Looper.getMainLooper());
            handler.postDelayed(new Runnable() {
                public void run() {
                    Logger.d(TAG, "Checking timeout for request(id=%s)", transaction.getRequest().getId());
                    handleResponse(ElectrodeBridgeResponse.createResponseForRequest(transaction.getRequest(), null, BridgeFailureMessage.create("EREQUESTTIMEOUT", "Request timeout")));
                }
            }, transaction.getRequest().getTimeoutMs());
        } else {
            Logger.d(TAG, "NO_TIMEOUT request, Will skip timeout check for request(%s)", transaction.getRequest());
        }
    }

    private void dispatchRequestToLocalHandler(@NonNull final BridgeTransaction transaction) {
        Logger.d(TAG, "Sending request(id=%s) to local handler", transaction.getRequest().getId());

        final ElectrodeBridgeRequest request = transaction.getRequest();
        sRequestDispatcher.dispatchRequest(transaction.getRequest(), new ElectrodeBridgeResponseListener<Object>() {
            @Override
            public void onFailure(@NonNull FailureMessage failureMessage) {
                ElectrodeBridgeResponse response = ElectrodeBridgeResponse.createResponseForRequest(request, null, failureMessage);
                handleResponse(response);
            }

            @Override
            public void onSuccess(@Nullable Object responseData) {
                ElectrodeBridgeResponse response = ElectrodeBridgeResponse.createResponseForRequest(request, responseData, null);
                handleResponse(response);
            }
        });

    }

    private void dispatchRequestToReact(@NonNull BridgeTransaction bridgeTransaction) {
        Logger.d(TAG, "Sending request(id=%s) over to JS side as there is no local request handler available", bridgeTransaction.getId());
        mReactContextWrapper.emitEvent(bridgeTransaction.getRequest());
    }

    private void handleResponse(@NonNull ElectrodeBridgeResponse bridgeResponse) {
        Logger.d(TAG, "Handling bridge response");
        BridgeTransaction transaction = sPendingTransactions.get(bridgeResponse.getId());
        if (transaction != null) {
            transaction.setResponse(bridgeResponse);
            completeTransaction(transaction);
        } else {
            Logger.i(TAG, "Response(id=%s, name=%s) will be ignored as the transaction for this request has already been removed from the queue. Perhaps it's already timed-out or completed", bridgeResponse.getId(), bridgeResponse.getName());
        }

    }

    private void notifyLocalEventListeners(@NonNull final ElectrodeBridgeEvent event) {
        mReactContextWrapper.runOnUiQueueThread(new Runnable() {
            @Override
            public void run() {
                sEventDispatcher.dispatchEvent(event);
            }
        });
    }

    private void notifyReactEventListeners(@NonNull ElectrodeBridgeEvent event) {
        mReactContextWrapper.emitEvent(event);
    }

    private void completeTransaction(@NonNull final BridgeTransaction transaction) {
        if (transaction.getResponse() == null) {
            throw new IllegalArgumentException("Cannot complete transaction, a transaction can only be completed with a valid response.");
        }
        Logger.d(TAG, "completing transaction(%s)", transaction.getId());

        sPendingTransactions.remove(transaction.getId());

        final ElectrodeBridgeResponse response = transaction.getResponse();
        logResponse(response);

        if (transaction.isJsInitiated()) {
            Logger.d(TAG, "Completing by emitting event to JS since the request was initiated from JS side.");
            mReactContextWrapper.emitEvent(response);
        } else {
            if (transaction.getFinalResponseListener() != null) {
                if (response.getFailureMessage() != null) {
                    Logger.d(TAG, "Completing by issuing a failure call back to local response listener.");
                    mReactContextWrapper.runOnUiQueueThread(new Runnable() {
                        @Override
                        public void run() {
                            transaction.getFinalResponseListener().onFailure(response.getFailureMessage());
                        }
                    });
                } else {
                    Logger.d(TAG, "Completing by issuing a success call back to local response listener.");
                    mReactContextWrapper.runOnUiQueueThread(new Runnable() {
                        @Override
                        public void run() {
                            transaction.getFinalResponseListener().onSuccess(response);
                        }
                    });
                }
            } else {
                throw new IllegalArgumentException("Should never reach here, a response listener should always be set for a local transaction");
            }

        }

    }

    private void logRequest(@NonNull ElectrodeBridgeRequest request) {
        Logger.d(TAG, "--> --> --> --> --> Request(id=%s, name=%s, isJS=%s)", request.getId(), request.getName(), request.isJsInitiated());
    }

    private void logResponse(ElectrodeBridgeResponse response) {
        Logger.d(TAG, "<-- <-- <-- <-- <-- Response(id=%s, name=%s, data=%s, error=%s) received", response.getId(), response.getName(), response.getData(), response.getFailureMessage());
    }

    public interface ReactNativeReadyListener {
        void onReactNativeReady();
    }

    private static ReactNativeReadyListener sReactNativeReadyListener;

    public static void registerReactNativeReadyListener(ReactNativeReadyListener listener) {
        // If react native initialization is already completed, just call listener
        // immediately
        if (sIsReactNativeReady) {
            listener.onReactNativeReady();
        }
        // Else it will get invoked whenever react native initialization is done
        else {
            sReactNativeReadyListener = listener;
        }
    }

    public void onReactNativeInitialized() {
        sIsReactNativeReady = true;
        if (sReactNativeReadyListener != null) {
            sReactNativeReadyListener.onReactNativeReady();
        }
    }

    /**
     * Method for testing only
     */
    @VisibleForTesting
    void debug_ClearRequestHandlerRegistrar() {
        ((RequestRegistrarImpl) sRequestRegistrar).reset();
    }
}
